### DISCUSSION TITLE
Flatted vs JSON

### DISCUSSION THREAD

**Discussion by samuelcastro** [UnlyEd/next-right-now#270] [Q&A] [answered]
On `MultiversalAppBootstrap` you were importing `customer` from `pageProps`, but now there is a `serializedDataset` which is  serialized using **Flatted**, to avoid contains circular references, I wonder the need to add another lib just to serialize the data considering that `Customer` (at least not on NRN) does not have a circular dependence.

I'm asking because this leaves the impression that NRN users should always use this approach for any object that they need to return on their static props SSG/SSR functions.

Have you considered to open up an issue on Next.js to suggest the usage of **Flatted** rather than `JSON.stringify`.

Thanks.

**Comment by Vadorequest** [comment, marked as answer]
> I wonder the need to add another lib just to serialize the data considering that Customer (at least not on NRN) does not have a circular dependence.

Honestly, it should be a default. If you return dynamic data from `getServerSideProps` or similar, and if those data contain circular reference, you'll break your whole page for your customers.

Serializing the data using a safe serializer (e.g: Flatted) has several advantages:
- Ensures it won't break because of dynamic data containing circular references
- Doesn't increase build size because executed on the server only
- Allows using circular references, which can be great when using nested entities (such as DB data models sent to the client)

I'm not sure the Next.js team would want to add it by default, but I strongly recommend it as soon as you forward dynamic data from the server to the client.

**Reply by samuelcastro** [reply]
I understand the benefits I just wonder why use it when there is no circular reference on `Customer`.

**Reply by Vadorequest** [reply]
Because it's a change made on the Airtable preset, where there are tons of circular references (maybe not in the preset, but definitely in our private fork).

So, I added it to NRN directly because I wasted days until I found the right/easiest way of doing things, using Flatted.

**Reply by Vadorequest** [reply]
So, it isn't useful immediately yeah, but as soon as you'll have a circular reference then it'll just work instead of crashing completely your app, I believe that's a better DX 😆  even though you might never even notice it helped you ahah

**Reply by samuelcastro** [reply]
Ok, I just opened up an issue there to see. https://github.com/vercel/next.js/issues/21364

**Reply by samuelcastro** [reply]
Btw, why are you sing [json-stringify-safe](https://www.npmjs.com/package/json-stringify-safe) and Flatted?

**Reply by Vadorequest** [reply]
They aren't used for the same thing, it could be replaced by Flatted, I guess. (but that'd be a breaking change for current `decodeQueryParameterToJSON` implementation)

https://github.com/UnlyEd/next-right-now/blob/0174dcc21e36dd356e2be475bfca5339baf27fd3/src/modules/core/js/url.ts#L16
